### PR TITLE
refactor aotautograd to set requires_grad on info rather than a separate array

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110773
* #110721
* __->__ #110720

### CODE REVIEW

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -3416,7 +3429,7 @@ def convert(idx, x):
             # aot_export: ban input metadata mutations for now to keep shared code paths simpler.
             # Keeping .resize_() in the graph will require some work
             # Allowing it but keeping the graph functional will require some calling convention changes.
-            if len([x for x in fw_metadata.input_info if x.mutates_metadata]) != 0:
+            if len([x for x in fw_metadata.requires_grad_info[:fw_metadata.num_mutated_inputs] if x]) != 0:
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:3432]: hmm, why? (this assert is only meant to be checking for input metadata mutations)

Comment (Chillee) [reply torch/_functorch/aot_autograd.py:3432]: I think this was an accident/merge issue

File: test/functorch/test_aotdispatch.py [hunk 2]
```diff
@@ -2095,6 +2095,7 @@ def forward(self, x):
 def extract_graph(fx_g, _, graph_cell):
     graph_cell[0] = fx_g
     return fx_g
+extract_graph._boxed_call = True
```
Comment (ezyang) [inline test/functorch/test_aotdispatch.py:2098]: How is this boxed 🧐

Comment (Chillee) [reply test/functorch/test_aotdispatch.py:2098]: Oh true.

File: test/functorch/test_aotdispatch.py [hunk 3]
```diff
@@ -317,7 +317,7 @@ def verify_aot_autograd(
             else:
                 compiled_f = aot_function(
                     f,
-                    fw_compiler=partial(extract_graph, graph_cell=fw_graph_cell),
+                    fw_compiler=make_boxed_compiler(partial(extract_graph, graph_cell=fw_graph_cell)),
```
Comment (bdhirsh) [inline test/functorch/test_aotdispatch.py:320]: thanks, these tests warnings have been pretty annoying

File: torch/_functorch/aot_autograd.py [hunk 4]
```diff
@@ -3114,11 +3133,19 @@ def forward(ctx, *deduped_flat_tensor_args):
             # invariant: intermediate bases always require gradients, so we don't have to
             # consider marking them as non-differentiable.
             raw_returns_not_including_intermediate_bases = raw_returns[:num_mutated_inputs + num_outputs]
+
+            raw_returns_meta = (
+                [x for x in CompiledFunction.metadata.input_info if x.mutates_data or x.mutates_metadata]
+                + CompiledFunction.metadata.output_info
+            )
+
+            for (i, x) in enumerate(raw_returns_not_including_intermediate_bases):
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:3142]: Maybe assert that `len(raw_returns_not_including_intermediate_bases) == len(raw_returns_meta)` too?

**Review by bdhirsh — APPROVED** [review]
thanks for the cleanup :)

**Pull request closed without merging**